lib/gltfPipeline.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+var fs = require('fs');
+var path = require('path');
+var addDefaults = require('./addDefaults');
+var removeUnusedImages = require('../').removeUnusedImages;
+var removeUnusedSamplers = require('../').removeUnusedSamplers;
+var removeUnusedShaders = require('../').removeUnusedShaders;
+var removeUnusedTechniques = require('../').removeUnusedTechniques;
+var removeUnusedPrograms = require('../').removeUnusedPrograms;
+var removeUnusedBuffers = require('../').removeUnusedBuffers;
+var removeUnusedBufferViews = require('../').removeUnusedBufferViews;
+var removeUnusedMaterials = require('../').removeUnusedMaterials;
+var removeUnusedSkins = require('../').removeUnusedSkins;
+var removeUnusedCameras = require('../').removeUnusedCameras;
+var removeUnusedTextures = require('../').removeUnusedTextures;
+var removeUnusedMeshes = require('../').removeUnusedMeshes;
+var removeUnusedNodes = require('../').removeUnusedNodes;
+var removeUnusedAccessors = require('../').removeUnusedAccessors;
+var removeUnused = require('./removeUnused');
+var loadGltfUris = require('./loadGltfUris');
+var parseBinaryGltf = require('./parseBinaryGltf');
+var addPipelineExtras = require('./addPipelineExtras');
+var convertDagToTree = require('./convertDagToTree');
+var combineMeshes = require('./combineMeshes');
+var combinePrimitives = require('./combinePrimitives');
+var removeUnusedVertices = require('./removeUnusedVertices');
+var OptimizationStatistics = require('./OptimizationStatistics');
+var Cesium = require('cesium');
+var defined = Cesium.defined;
+var defaultValue = Cesium.defaultValue;
+var DeveloperError = Cesium.DeveloperError;
+
+
+module.exports = gltfPipeline;
+
+function gltfPipeline(options, callback) {
+    var gltfPath = options.inputPath;
+    
+    if (!defined(gltfPath)) {
+        throw new DeveloperError('Input path is undefined');
+    }
+    
+    var fileExtension = path.extname(gltfPath);
+    var fileName = path.basename(gltfPath, fileExtension);
+    var filePath = path.dirname(gltfPath);
+
+    fs.readFile(gltfPath, function (err, data) {
+        if (err) {
+            throw err;
+        }
+
+        var gltf;
+        if (fileExtension === '.glb') {
+            gltf = parseBinaryGltf(data);
+        }
+        else if (fileExtension === '.gltf') {
+            gltf = JSON.parse(data);
+            addPipelineExtras(gltf);
```

Comment:
We moved `addPipelineExtras` here from further down to resolve a bug where the gltfPipeline couldn't load glbs with the `KHR_binary_glTF extension`. It looks like glbs with that extension have an empty data URI, which isn't a problem for the pipeline if the data that should be there is in the assets' `extra` attribute. `parseBinaryGltf` was handling this, but then the later call to `addPipelineExtras` seemed to wipe out the `extra` attributes, causing an error later down in `loadGltfUris`.